### PR TITLE
Framework: add separate server state hydration

### PR DESCRIPTION
**Pull request by gwwar**
Following the conversation in #3337 This is an example of what a separate server state hydration pipeline might look like. Reducers would only need to implement this for subtrees that do not contain plain JavaScript objects. Feel free to call out other suggestions for the action name.

cc @aduth @seear

### CODE REVIEW

File: client/state/initial-state.js [hunk 1]
```diff
@@ -24,7 +25,8 @@ export const MAX_AGE = 7 * DAY_IN_HOURS * HOUR_IN_MS;
 function getInitialServerState() {
 	// Bootstrapped state from a server-render
 	if ( typeof window === 'object' && window.initialReduxState ) {
-		return window.initialReduxState;
+		const serverState = reducer( window.initialReduxState, { type: SERVER_DESERIALIZE } );
+		return pick( serverState, Object.keys( window.initialReduxState ) );
```
Comment (aduth) [inline client/state/initial-state.js]: Why do you need to `pick` here? Related, if there's a need to trim reduced state to reflect that of the global object, is there not a concern about nested structures?

Comment (gwwar) [inline client/state/initial-state.js]: If I don't, server state will override persisted state with default initial state.

@seear which sections of the tree do you see us bootstrapping, and do we need to be more precise?

File: client/state/initial-state.js [hunk 2]
```diff
@@ -2,12 +2,13 @@
  * External dependencies
  */
 import debugModule from 'debug';
+import pick from 'lodash/object/pick';
```
Comment (blowery) [inline client/state/initial-state.js]: lodash 4 compat warning

Comment (gwwar) [inline client/state/initial-state.js]: :+1: rebased and updated in 1f9e867

File: client/state/initial-state.js (hunk 1 above)
Comment (seear) [inline client/state/initial-state.js]: Currently the state tree being bootstrapped is defined by the [`pick` on the server](https://github.com/Automattic/wp-calypso/blob/bcfe9f49d801818e56bfa87ea8dce5709dff88b9/server/pages/index.js#L382). This works fine for now, but perhaps once the server code becomes more modularized, we can remove the server `pick` and change this client code to use a formally-defined bootstrap tree.

A reducer that did not return any default state for the `SERVER_DESERIALIZE` action type would be great for that. Not sure if that is possible.

File: client/state/action-types.js [hunk 3]
```diff
@@ -50,6 +50,7 @@ export const RECEIPT_FETCH_FAILED = 'RECEIPT_FETCH_FAILED';
 export const REMOVE_NOTICE = 'REMOVE_NOTICE';
 export const SELECTED_SITE_SET = 'SELECTED_SITE_SET';
 export const SERIALIZE = 'SERIALIZE';
+export const SERVER_DESERIALIZE = 'SERVER_DESERIALIZE';
```
Comment (seear) [inline client/state/action-types.js]: I was going to say that `BOOTSTRAP` may give more clue as to what this is for, but I checked back, and that term is not used at all in the server-rendering section of the [Redux docs](http://redux.js.org/docs/recipes/ServerRendering.html), so I think it is fine as is.

**Comment by seear** [comment]
This is working fine for the server state currently in usage. Good to merge :+1:

**Comment by gwwar** [comment]
Thanks for the reviews @seear & @aduth. Going to squash commits.

File: client/state/action-types.js (hunk 3 above)
Comment (gwwar) [inline client/state/action-types.js]: I'm not attached to the name, so I'd be happy to open up another PR if we find that another term is more clear.
